Userland/Libraries/LibGUI/TextDocument.cpp

```diff
@@ -383,6 +407,96 @@ DeprecatedString TextDocument::text_in_range(TextRange const& a_range) const
     return builder.to_deprecated_string();
 }
 
+// This function will determine the number of codepoints of the glyph immediately preceding the cursor, by determining if there is an emoji. If no emoji is found it will return 1.
+size_t TextDocument::get_code_points_before_cursor(TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    return get_prev_grapheme_cluster_boundary(cursor);
+}
+
+// This function will return the position of the next grapheme cluster break, relative to the cursor, for "correct looking" parsing of unicode based on grapheme cluster boundary algorithm.
+size_t TextDocument::get_next_grapheme_cluster_boundary(TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line());
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+
+    if (line_span.is_empty())
+        return 0;
+
+    if (cursor.column() == line.length())
+        return 0;
+
+    auto boundaries_cache = Unicode::find_grapheme_segmentation_boundaries(Utf32View(line.code_points(), line.length()));
+
+    if (boundaries_cache.is_empty())
+        return cursor.column();
+
+    if (boundaries_cache.size() == 1)
+        return cursor.column() - boundaries_cache.at(0);
```

Comment:
Did you have a case where this is ever true? If `find_grapheme_segmentation_boundaries` appends anything, it appends at least 2 indices (the beginning and end of the view it's given).

Reply:
No. I've removed it.